controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -343,23 +360,136 @@ public void deleteStream(StreamInfo request, StreamObserver<DeleteStreamStatus>
+                        request.getScope(), request.getStream());
+
+                final AuthHandler.Permissions authAndTokenPermission;
+                if (authParams.isMarkStream()) {
+                    // Clients are allowed to read from a mark stream, but aren't allowed to write to it. Since the
+                    // client didn't specify the access operation, we assume here that it intends to read from it.
+                    authAndTokenPermission = AuthHandler.Permissions.READ;
+                } else {
+                    authAndTokenPermission = AuthHandler.Permissions.READ_UPDATE;
+                }
+                this.grpcAuthHelper.checkAuthorization(resource, authAndTokenPermission);
+                return this.grpcAuthHelper.createDelegationToken(streamResource, authAndTokenPermission);
+            } else {
+                log.trace("Access operation was {} for request with scope {} and stream {}",
+                        translate(request.getAccessOperation()), request.getScope(), request.getStream());
+
+                // The resource string that'll be used in the delegation token for use of the segment store
+                final String tokenResource;
+
+                // The operation that'll be specified as granted for the resource in the token. The bearer of the token
+                // will be allowed to perform the specified operation.
+                final AuthHandler.Permissions tokenPermission;
+
+                // This is the permission that the client is requesting to be assigned on the delegation token.
+                AuthHandler.Permissions requestedPermissions = authParams.requestedPermission();
+
+                if (authParams.isStreamUserDefined()) {
+                    // The operation itself requires the caller to possess read permissions.
+                    AuthHandler.Permissions minimumPermissions = AuthHandler.Permissions.READ;
+
+                    if (requestedPermissions.equals(AuthHandler.Permissions.READ_UPDATE) ||
+                            requestedPermissions.equals(minimumPermissions)) {
+                        this.grpcAuthHelper.checkAuthorization(streamResource, requestedPermissions);
+                        tokenResource = streamResource;
+                        tokenPermission = requestedPermissions;
+                    } else {
+                        // The minimum permission that the user must have to be able to invoke this call. This
+                        // authorizes the operation.
+                        this.grpcAuthHelper.checkAuthorization(streamResource, minimumPermissions);
+
+                        // Here, we check whether the user is authorized for the requested access.
+                        this.grpcAuthHelper.checkAuthorization(streamResource, requestedPermissions);
+                        tokenResource = streamResource;
+                        tokenPermission = requestedPermissions;
+                    }
+                } else {
```

Comment:
This else block starting at line 464 is about internal streams. Are the tokens produced here both for the use of application clients and the controller? For application clients, is there a case in which we want to authorize them to read from internal streams that are not mark streams,(e.g., streams the controller use with its workflows)? Is it for some debug or recovery scenario?

Reply:
> ... Are the tokens produced here both for the use of application clients and the controller?

No, only application clients making the gRPC call will arrive here. The Controller itself uses `LocalController` as I mentioned in the previous comment.

Reply:
> For application clients, is there a case in which we want to authorize them to read from internal streams that are not mark streams,(e.g., streams the controller use with its workflows)? Is it for some debug or recovery scenario?

The reader group stream is an example of internal stream that application clients may access (usually via the client library). I don't know the answer for other types of internal streams, but the older code allowed for it.

Does that answer your question?

Reply:
Yes, that's right.